File: func-sockets/clients/graph-socket.js
/**
 * Browser client for one graph room on a Func Sockets relay.
 *
 * Use this file in the UI to send graph commands and receive backend events.
 * It handles connection and JSON decoding only; it does not update UI state or
 * interpret the command/event language.
 *
 * @example
 * const socket = new GraphSocket('graph-42')
 * socket.addEventListener('message', (event) => graphStore.apply(event.detail))
 * socket.connect()
 */
class GraphSocket extends EventTarget {
    /**
     * Prepare a graph client without opening its connection.
     *
     * @param {string} graphId - Graph room to join.
     * @param {string} [baseUrl='ws://127.0.0.1:8777'] - Relay server URL.
     *
     * @example
     * const socket = new GraphSocket('demo', 'ws://127.0.0.1:8777')
     */
    constructor(graphId, baseUrl = 'ws://127.0.0.1:8777') {
        super()
        this.graphId = graphId
        this.baseUrl = baseUrl.replace(/\/$/, '')
        this.socket = null
    }

    /**
     * Open the WebSocket and forward its lifecycle and message events.
     *
     * Listen for `open`, `close`, `error`, and `message` on this GraphSocket.
     * JSON messages are decoded into `event.detail`; other text and binary
     * values are passed through as received.
     *
     * @returns {WebSocket} The browser WebSocket that was opened.
     *
     * @example
     * socket.addEventListener('open', () => console.log('ready'))
     * socket.connect()
     */
    connect() {
        const graphId = encodeURIComponent(this.graphId)
        this.socket = new WebSocket(`${this.baseUrl}/graph/${graphId}`)
        this.socket.addEventListener('open', (event) => this.dispatchEvent(new Event('open')))
        this.socket.addEventListener('close', (event) => this.dispatchEvent(new Event('close')))
        this.socket.addEventListener('error', (event) => this.dispatchEvent(new Event('error')))
        this.socket.addEventListener('message', (event) => {
            let value = event.data
            if (typeof value === 'string') {
                try {
                    value = JSON.parse(value)
                } catch (error) {
                    // Non-JSON text is valid relay traffic.
                }
            }
            this.dispatchEvent(new CustomEvent('message', { detail: value }))
        })
        return this.socket
    }

    /**
     * Send a command or other application message to peers in this graph.
     *
     * Strings are sent unchanged. Other values are encoded as JSON. Call this
     * only after the `open` event.
     *
     * @param {string|object} message - Payload to relay.
     *
     * @example
     * socket.send({ kind: 'command', name: 'graph.snapshot.get' })
     */
    send(message) {
        if (!this.socket || this.socket.readyState !== WebSocket.OPEN) {
            throw new Error('GraphSocket is not connected')
        }
        this.socket.send(typeof message === 'string' ? message : JSON.stringify(message))
    }

    /**
     * Move the existing connection to another graph room.
     *
     * The relay replies with a normal `message` event containing
     * `{ type: 'bound', graph_id: graphId }`.
     *
     * @param {string} graphId - New graph room to join.
     *
     * @example
     * socket.bind('graph-43')
     */
    bind(graphId) {
        this.graphId = graphId
        this.send({ type: 'bind', graph_id: graphId })
    }

    /**
     * Close the underlying WebSocket if it exists.
     *
     * @example
     * socket.close()
     */
    close() {
        this.socket?.close()
    }
}

File: func-sockets/clients/graph_socket.py
"""Reusable Python client for a single Func Sockets graph room.

Use :class:`GraphSocket` in the Python graph process to publish events and
receive UI commands without handling WebSocket setup in graph code.

Minimal use::

    async with GraphSocket("graph-42") as socket:
        await socket.send({"kind": "event", "name": "graph.idle"})
        command = await socket.receive()
"""

from __future__ import annotations

import json
from typing import Any
from urllib.parse import quote

from websockets.asyncio.client import connect
from websockets.exceptions import ConnectionClosed


class GraphSocket:
    """Connect a Python producer or consumer to one graph room.

    The context manager opens the connection, waits for the server's binding
    confirmation, and closes cleanly when the block exits.
    """

    def __init__(self, graph_id: str, base_url: str = "ws://127.0.0.1:8777") -> None:
        """Prepare a client URL for ``graph_id`` without connecting yet.

        Example::

            socket = GraphSocket("demo", "ws://127.0.0.1:8777")
        """
        self.url = f"{base_url.rstrip('/')}/graph/{quote(graph_id, safe='')}"
        self.websocket = None

    async def __aenter__(self) -> GraphSocket:
        """Open the socket and return this bound client.

        Prefer ``async with GraphSocket("demo") as socket`` over calling this
        method directly.
        """
        self.websocket = await connect(self.url)
        await self.websocket.recv()
        return self

    async def __aexit__(self, *exc_info: object) -> None:
        """Close the socket when leaving its ``async with`` block."""
        if self.websocket is not None:
            await self.websocket.close()
            self.websocket = None

    async def send(self, message: str | bytes | dict[str, Any]) -> None:
        """Send text, bytes, or a JSON-serializable dictionary to graph peers.

        Example::

            await socket.send({"kind": "event", "name": "node.updated"})
        """
        if self.websocket is None:
            raise RuntimeError("GraphSocket is not connected")
        if isinstance(message, dict):
            message = json.dumps(message)
        await self.websocket.send(message)

    async def receive(self) -> str | bytes:
        """Wait for and return the next message from another graph peer.

        JSON remains encoded text so the graph can choose its own decoder.

        Example::

            raw_command = await socket.receive()
        """
        if self.websocket is None:
            raise RuntimeError("GraphSocket is not connected")
        return await self.websocket.recv()

    def __aiter__(self) -> GraphSocket:
        """Return this client as an async stream of incoming messages."""
        return self

    async def __anext__(self) -> str | bytes:
        """Return the next message, ending iteration after socket closure.

        Example::

            async for message in socket:
                handle(message)
        """
        if self.websocket is None:
            raise StopAsyncIteration
        try:
            return await self.websocket.recv()
        except ConnectionClosed as exc:
            raise StopAsyncIteration from exc
